data_prep/graph_dataset.py

```diff
@@ -1,72 +1,155 @@
 import abc
 
-from data_prep.dataset import Dataset
+import torch
 from torch_geometric.data import Dataset as GeometricDataset
+from torch_geometric.data import Data
+from transformers import RobertaTokenizerFast
+
+from data_prep.dataset import Dataset
+from data_prep.graph_utils import tf_idf_mtx, get_PMI
 
 
 class GraphDataset(Dataset, GeometricDataset):
     """
     Parent class for graph datasets.
     Require to implement a preprocess and generate_features methods.
     """
-    def __init__(self, *args, **kwargs):
-        # preprocess text
-        # compute tf.idf and pmi
-        # generate edges
-        # generate features
-        # build data object
-        raise NotImplementedError
+    def __init__(self, corpus, device):
+        self._device = device
+        self._tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
+        self._num_classes = corpus.num_classes()
+
+        train_texts, train_labels = corpus.train_data()
+        val_texts, val_labels = corpus.val_data()
+        test_texts, test_labels = corpus.test_data()
+
+        self._raw_texts = train_texts + val_texts + test_texts
+        # Hopefully no tokenizer makes a token "doc.i"
+        all_docs = ['doc.{}'.format(i) for i in range(len(self._raw_texts))]
+
+        print('Preprocess corpus')
+        tokenized_text, self._tokens = self._preprocess(self._raw_texts)
+
+        iton = list(all_docs + self._tokens)
+        ntoi = {iton[i]: i for i in range(len(iton))}
+
+        print('Compute tf.idf')
+        tf_idf, tf_idf_words = tf_idf_mtx(tokenized_text)
+
+        print('Compute PMI scores')
+        pmi_score = get_PMI(tokenized_text)
+
+        print('Generate edges')
+        edge_index, edge_attr = self._generate_edges(tf_idf, tf_idf_words, pmi_score, ntoi)
+
+        print('Generate masks')
+        train_mask, val_mask, test_mask = self._generate_masks(len(train_texts), len(train_texts),
+                                                              len(train_texts), len(iton))
+
+        doc_labels = train_labels + val_labels + test_labels
+        self._labels = torch.zeros(len(iton), device=self._device) - 1
+        self._labels[:len(doc_labels)] = torch.tensor(doc_labels)
+
+        self._data = Data(edge_index=edge_index, edge_attr=edge_attr, y=self._labels)
+        self._data.train_mask = train_mask
+        self._data.val_mask = val_mask
+        self._data.test_mask = test_mask
 
+    # THIS FUNCTION IS NOT REALLY NEEDED
```

Comment:
Why? I thought it would be used on the `train.get_dataloaders` to get the number of classes for the linear classification layer

Reply:
This depends on whether at the time when the classification layer needs the number of classes we have access to the `Data` object. If yes, then we can just ask `num_classes` from there and this function is redundant. Otherwise it is needed.